File: backend/app/models/purchase.py
"""Purchase workflow models."""
from uuid import uuid4
from datetime import datetime
from sqlalchemy import Column, String, Integer, Boolean, DateTime, Date, UUID, ForeignKey, Numeric, Text
from app.database import Base


class PurchaseOrder(Base):
    __tablename__ = "purchase_orders"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    po_number = Column(String(30), unique=True, nullable=False)
    supplier_id = Column(UUID(as_uuid=True), ForeignKey("suppliers.id"), nullable=False)
    order_date = Column(Date, nullable=False)
    expected_delivery_date = Column(Date, nullable=True)
    status = Column(String(20), nullable=False, default="draft")
    subtotal = Column(Integer, nullable=False, default=0)
    total_discount = Column(Integer, nullable=False, default=0)
    total_taxable_amount = Column(Integer, nullable=False, default=0)
    total_cgst = Column(Integer, nullable=False, default=0)
    total_sgst = Column(Integer, nullable=False, default=0)
    total_igst = Column(Integer, nullable=False, default=0)
    total_gst = Column(Integer, nullable=False, default=0)
    total_amount = Column(Integer, nullable=False, default=0)
    notes = Column(Text, nullable=True)
    is_deleted = Column(Boolean, nullable=False, default=False)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)


class PurchaseOrderItem(Base):
    __tablename__ = "purchase_order_items"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    purchase_order_id = Column(UUID(as_uuid=True), ForeignKey("purchase_orders.id"), nullable=False)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=False)
    description = Column(String(255), nullable=True)
    quantity = Column(Numeric(12, 4), nullable=False)
    unit_price = Column(Integer, nullable=False)
    discount_percent = Column(Numeric(5, 2), nullable=False, default=0)
    discount_amount = Column(Integer, nullable=False, default=0)
    taxable_amount = Column(Integer, nullable=False)
    gst_rate = Column(Integer, nullable=False)
    cgst_amount = Column(Integer, nullable=False, default=0)
    sgst_amount = Column(Integer, nullable=False, default=0)
    igst_amount = Column(Integer, nullable=False, default=0)
    total_amount = Column(Integer, nullable=False)
    received_quantity = Column(Numeric(12, 4), nullable=False, default=0)
    is_deleted = Column(Boolean, nullable=False, default=False)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)


class GoodsReceiptNote(Base):
    __tablename__ = "goods_receipt_notes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    grn_number = Column(String(30), unique=True, nullable=False)
    purchase_order_id = Column(UUID(as_uuid=True), ForeignKey("purchase_orders.id"), nullable=True)
    supplier_id = Column(UUID(as_uuid=True), ForeignKey("suppliers.id"), nullable=False)
    supplier_invoice_number = Column(String(50), nullable=True)
    supplier_invoice_date = Column(Date, nullable=True)
    receipt_date = Column(Date, nullable=False)
    status = Column(String(20), nullable=False, default="draft")
    subtotal = Column(Integer, nullable=False, default=0)
    total_discount = Column(Integer, nullable=False, default=0)
    total_taxable_amount = Column(Integer, nullable=False, default=0)
    total_cgst = Column(Integer, nullable=False, default=0)
    total_sgst = Column(Integer, nullable=False, default=0)
    total_igst = Column(Integer, nullable=False, default=0)
    total_gst = Column(Integer, nullable=False, default=0)
    total_amount = Column(Integer, nullable=False, default=0)
    notes = Column(Text, nullable=True)
    is_deleted = Column(Boolean, nullable=False, default=False)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)


class GRNItem(Base):
    __tablename__ = "grn_items"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    grn_id = Column(UUID(as_uuid=True), ForeignKey("goods_receipt_notes.id"), nullable=False)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=False)
    purchase_order_item_id = Column(UUID(as_uuid=True), ForeignKey("purchase_order_items.id"), nullable=True)
    quantity = Column(Numeric(12, 4), nullable=False)
    unit_price = Column(Integer, nullable=False)
    discount_percent = Column(Numeric(5, 2), nullable=False, default=0)
    discount_amount = Column(Integer, nullable=False, default=0)
    taxable_amount = Column(Integer, nullable=False)
    gst_rate = Column(Integer, nullable=False)
    cgst_amount = Column(Integer, nullable=False, default=0)
    sgst_amount = Column(Integer, nullable=False, default=0)
    igst_amount = Column(Integer, nullable=False, default=0)
    total_amount = Column(Integer, nullable=False)
    is_deleted = Column(Boolean, nullable=False, default=False)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)


class PurchaseReturn(Base):
    __tablename__ = "purchase_returns"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    return_number = Column(String(30), unique=True, nullable=False)
    supplier_id = Column(UUID(as_uuid=True), ForeignKey("suppliers.id"), nullable=False)
    grn_id = Column(UUID(as_uuid=True), ForeignKey("goods_receipt_notes.id"), nullable=False)
    return_date = Column(Date, nullable=False)
    reason = Column(Text, nullable=False)
    status = Column(String(20), nullable=False, default="draft")
    subtotal = Column(Integer, nullable=False, default=0)
    total_gst = Column(Integer, nullable=False, default=0)
    total_amount = Column(Integer, nullable=False, default=0)
    is_deleted = Column(Boolean, nullable=False, default=False)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)


class PurchaseReturnItem(Base):
    __tablename__ = "purchase_return_items"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    purchase_return_id = Column(UUID(as_uuid=True), ForeignKey("purchase_returns.id"), nullable=False)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=False)
    grn_item_id = Column(UUID(as_uuid=True), ForeignKey("grn_items.id"), nullable=True)
    quantity = Column(Numeric(12, 4), nullable=False)
    unit_price = Column(Integer, nullable=False)
    taxable_amount = Column(Integer, nullable=False)
    gst_rate = Column(Integer, nullable=False)
    cgst_amount = Column(Integer, nullable=False, default=0)
    sgst_amount = Column(Integer, nullable=False, default=0)
    igst_amount = Column(Integer, nullable=False, default=0)
    total_amount = Column(Integer, nullable=False)
    is_deleted = Column(Boolean, nullable=False, default=False)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)

File: frontend/src/api/purchase.ts
/**
 * Purchase API Client
 * 
 * Handles all purchase workflow operations:
 * - Purchase Orders (PO)
 * - Goods Receipt Notes (GRN)
 * - Purchase Returns
 */

import { apiClient } from './client';

// Request payload types (matching backend schemas)
export interface PurchaseLineItem {
  id?: string;
  product_id: string;
  description?: string;
  quantity: number;
  unit_price: number;
  discount_percent?: number;
  gst_rate: number;
  purchase_order_item_id?: string;
  received_quantity?: number;
}

export interface CreatePOPayload {
  supplier_id: string;
  order_date: string;
  expected_delivery_date?: string;
  notes?: string;
  status?: string;
  currency_code?: string;
  exchange_rate?: number;
  items: PurchaseLineItem[];
}

export interface UpdatePOPayload extends Partial<CreatePOPayload> {}

export interface UpdatePOStatusPayload {
  status: string;
}

export interface CreateGRNPayload {
  supplier_id: string;
  purchase_order_id?: string;
  supplier_invoice_number?: string;
  supplier_invoice_date?: string;
  receipt_date: string;
  notes?: string;
  items: PurchaseLineItem[];
}

export interface UpdateGRNPayload extends Partial<CreateGRNPayload> {}

export interface PurchaseReturnLineItem {
  product_id: string;
  grn_item_id?: string;
  quantity: number;
  unit_price: number;
  gst_rate: number;
}

export interface CreatePurchaseReturnPayload {
  supplier_id: string;
  grn_id: string;
  return_date: string;
  reason: string;
  items: PurchaseReturnLineItem[];
}

// Response types (from backend models)
export interface PurchaseOrder {
  id: string;
  po_number: string;
  supplier_id: string;
  order_date: string;
  expected_delivery_date?: string;
  status: string;
  subtotal: number;
  total_discount: number;
  total_taxable_amount: number;
  total_cgst: number;
  total_sgst: number;
  total_igst: number;
  total_gst: number;
  total_amount: number;
  currency_code?: string;
  exchange_rate?: number;
  notes?: string;
  created_at: string;
  created_by?: string;
}

export interface PurchaseOrderDetail {
  purchase_order: PurchaseOrder;
  items: PurchaseLineItem[];
}

export interface GoodsReceiptNote {
  id: string;
  grn_number: string;
  purchase_order_id?: string;
  supplier_id: string;
  supplier_invoice_number?: string;
  supplier_invoice_date?: string;
  receipt_date: string;
  status: string;
  subtotal: number;
  total_discount: number;
  total_taxable_amount: number;
  total_cgst: number;
  total_sgst: number;
  total_igst: number;
  total_gst: number;
  total_amount: number;
  notes?: string;
  created_at: string;
}

export interface GRNDetail {
  grn: GoodsReceiptNote;
  items: GRNItemResponse[];
}

export interface GRNItemResponse {
  id: string;
  grn_id: string;
  product_id: string;
  purchase_order_item_id?: string;
  quantity: number;
  unit_price: number;
  discount_percent: number;
  discount_amount: number;
  taxable_amount: number;
  gst_rate: number;
  cgst_amount: number;
  sgst_amount: number;
  igst_amount: number;
  total_amount: number;
}

export interface PurchaseReturn {
  id: string;
  return_number: string;
  supplier_id: string;
  grn_id: string;
  return_date: string;
  reason: string;
  status: string;
  subtotal: number;
  total_gst: number;
  total_amount: number;
  created_at: string;
}

class PurchaseApiClient {
  // ========== Purchase Orders ==========

  async listPOs(status?: string, page = 1, page_size = 20) {
    const params: Record<string, string | number> = { page, page_size };
    if (status) params.status = status;
    return apiClient.get<{ items: PurchaseOrder[]; total: number }>('/api/v1/purchase-orders', { params });
  }

  async getPO(id: string) {
    return apiClient.get<PurchaseOrderDetail>(`/api/v1/purchase-orders/${id}`);
  }

  async createPO(payload: CreatePOPayload) {
    return apiClient.post<PurchaseOrder>('/api/v1/purchase-orders', payload);
  }

  async updatePO(id: string, payload: UpdatePOPayload) {
    return apiClient.put<PurchaseOrder>(`/api/v1/purchase-orders/${id}`, payload);
  }

  async updatePOStatus(id: string, status: string) {
    return apiClient.patch<PurchaseOrder>(`/api/v1/purchase-orders/${id}/status`, { status });
  }

  // ========== Goods Receipt Notes ==========

  async listGRNs(status?: string, page = 1, page_size = 20) {
    const params: Record<string, string | number> = { page, page_size };
    if (status) params.status = status;
    return apiClient.get<{ items: GoodsReceiptNote[]; total: number }>('/api/v1/grn', { params });
  }

  async getGRN(id: string) {
    return apiClient.get<GRNDetail>(`/api/v1/grn/${id}`);
  }

  async createGRN(payload: CreateGRNPayload) {
    return apiClient.post<GoodsReceiptNote>('/api/v1/grn', payload);
  }

  async updateGRN(id: string, payload: UpdateGRNPayload) {
    return apiClient.put<GoodsReceiptNote>(`/api/v1/grn/${id}`, payload);
  }

  async confirmGRN(id: string) {
    return apiClient.post<GoodsReceiptNote>(`/api/v1/grn/${id}/confirm`, {});
  }

  async cancelGRN(id: string) {
    return apiClient.post<GoodsReceiptNote>(`/api/v1/grn/${id}/cancel`, {});
  }

  // ========== Purchase Returns ==========

  async listPurchaseReturns(page = 1, page_size = 20) {
    return apiClient.get<{ items: PurchaseReturn[]; total: number }>('/api/v1/purchase-returns', {
      params: { page, page_size },
    });
  }

  async getPurchaseReturn(id: string) {
    return apiClient.get<PurchaseReturn>(`/api/v1/purchase-returns/${id}`);
  }

  async createPurchaseReturn(payload: CreatePurchaseReturnPayload) {
    return apiClient.post<PurchaseReturn>('/api/v1/purchase-returns', payload);
  }

  async confirmPurchaseReturn(id: string) {
    return apiClient.post<PurchaseReturn>(`/api/v1/purchase-returns/${id}/confirm`, {});
  }

  async cancelPurchaseReturn(id: string) {
    return apiClient.post<PurchaseReturn>(`/api/v1/purchase-returns/${id}/cancel`, {});
  }

  // ========== PDF Download ==========

  async downloadPOPdf(id: string) {
    return apiClient.get(`/api/v1/purchase-orders/${id}/pdf`, {
      responseType: 'blob',
    });
  }
}

export const purchaseApi = new PurchaseApiClient();
